CONTAINER_R_CMD="docker run -v /:/home/user -w /home/user"+os.getcwd()+" fgajardoe/r-custom:latest "
#CONTAINER_R_CMD="docker run -v /:/home/user -w /home/user/storage2/dump/Orestias/TE-workflows/genome_size_contrib fgajardoe/r-custom:latest "


#configfile: "config.common.pairwise.yaml"
configfile: "config.common.yaml"

rule calcContrib:
	input:
		config["rmgr"]

	output:
		config["prefix"]+".RData"
	params:
		config["ordersToConsider"],
		config["prefix"]

	shell:
		CONTAINER_R_CMD+" Rscript calcContrib.R {input} {params}"

rule mergeContribTables:
	input:
		config["tree"],
		expand("{sp}.gene-annotation.gff",sp=config["species"]),
		expand("{sp}.genomeSize.tab",sp=config["species"]),
		expand("{sp}.RData",sp=config["species"])
	output:
		config["mergePrefix"]+".pdf",
		config["mergePrefix"]+".logfc.RData"	
	params:
		config["mergePrefix"],
		config["collapse_lower_than_mean_contrib"],
		config["onlyCoreTEsuperfams"],
		config["TEordersToCollapseAsOther"],
		config["refSpecieForLog2FC"],
		config["speciesColourFile"],
		config["forceThisYlimit"]
	shell:
		CONTAINER_R_CMD+" Rscript mergeMultispeciesContrib.R {params} {input}"


rule merge_LogFC_and_kimura_plots:
	input:
		config["mergePrefix"]+".logfc.RData",	
		config["kimuraPlotsRData"]
	output:
		config["mergePrefix"]+".LogFC_kimura_plots.pdf"
	shell:
		CONTAINER_R_CMD+" Rscript merge_LogFC_and_kimura_plots.R {input} "+config["mergePrefix"]

